modules/videoio: add libcamera backend for OpenCV VideoCapture

ssxc16 opened the pull request:
### Feature: Add Libcamera backend

**Summary:**

This pull request introduces a new backend for the `cv::VideoCapture` module, providing support for cameras using the Libcamera library on Linux (and potentially other supported platforms).

You can find the complete implementation [here](https://github.com/QueenofUSSR/opencv-RPiCamera/tree/lk_temp).
Detailed usage instructions and sample testing program can be found [here](https://gist.github.com/ssxc16/6b4f6d1fa1a9f72ddefb1bb00ee13aaf).

**Motivation / Reason for this change:**

Libcamera is the modern standard API for camera control on many recent Linux systems, including popular development boards like the Raspberry Pi. It offers a more robust and feature-rich interface compared to the legacy V4L2 API for many newer camera modules. Adding official Libcamera support to OpenCV will allow users on these platforms to fully utilize their cameras with `cv::VideoCapture`, replacing or complementing the existing V4L2 backend where applicable.

**Implementation Details:**

*   The backend is implemented as a plugin, allowing for dynamic building. This makes the Libcamera library an optional dependency for the main OpenCV build.
*   A new backend implementation `cap_libcamera.cpp` has been added to the `modules/videoio/src/` directory.
*   The new backend registers itself under the API ID `cv::VideoCaptureAPIs::CAP_LIBCAMERA`.
*   The implementation utilizes the Libcamera C++ API to enumerate cameras, configure streams, capture frames, and set/get basic properties.
*   **Reference to LCCV:** The design and implementation concepts are significantly inspired by and borrow structure from the excellent [LCCV project](https://github.com/kbarni/LCCV). 

**Implemented Features:**

*   `grabFrame`, `retrieveFrame` for frame capture.
*   Setting and getting basic frame properties (width, height, frame rate - `CAP_PROP_FRAME_WIDTH`, `CAP_PROP_FRAME_HEIGHT`, `CAP_PROP_FPS` eg).
*   Fast Frame Reading and Writing

**Issues of multi-cameras**

- [x] multiple USB cameras (with other backends)
- [x] one USB camera and one MIPI camera (index 0 will be the first recognized camera)
- [ ] multiple MIPI cameras **[WIP]**

QueenofUSSR committed:
Sync with 4.x

QueenofUSSR committed:
Fix paste error

QueenofUSSR committed:
Fix deadlock when cap.release() is called

QueenofUSSR committed:
Modify code formatting

ssxc16 committed:
Merge branch '4.x' into 4.x

ssxc16 committed:
Convert tabs to spaces

QueenofUSSR committed:
Deprecate unnecessary video thread function and buffer

Lookplanes committed:
(not tested yet)Remove videothread and turn to producer-consumer queue.

Lookplanes committed:
Merge branch '4.x' of github.com:QueenofUSSR/opencv-RPiCamera into lk_temp

Lookplanes committed:
Fix cap_libcamera_lk and replace the origin files; Add test files for libcamera

Lookplanes committed:
fix: Resolve libcamera compilation errors and enhance testing infrastructure

Lookplanes committed:
fix detect

Lookplanes committed:
Support single submit(like no queue)

Lookplanes committed:
Add C style api for plugin compilation

Lookplanes committed:
Less copies with own allocator

Lookplanes committed:
Simplify some code and suppress unused parameter warnings

ssxc16 committed:
Merge branch 'opencv:4.x' into lk_temp

opencv-alalek commented on modules/videoio/src/cap_libcamera_del.hpp:
```diff
@@ -0,0 +1,396 @@
+#pragma once
```
> _del

They are still here.

opencv-alalek reviewed (commented):
See also comments in the previous PR: #27254

Lookplanes committed:
Fix CMake and namespace format and add opencv license header

Lookplanes committed:
chore: Remove local test directory from version control

Lookplanes committed:
ci: Ignore local test directory

Lookplanes committed:
Use Producer-Consumer Queue implementation

QueenofUSSR committed:
Add OpenCV license header and fix code format

Lookplanes committed:
Use only on demand mode. Remove P-C Queue

Commit:
Refactor LibcameraCapture and fix 'failed to open camera by specialized index'

asmorkalov commented on .gitignore:
```diff
@@ -25,3 +25,21 @@ build
 xcuserdata/
+
+# Compiled executables and test files
+test/detect_backend
+test/test_libcamera
+test/test_libcamera2
+test/test_libcamera_plugin
+test/*.jpg
+test/*.png
+test/frame/
+test/test_results/
+test/libcamera_test_results/
+test/build_plugin/
```
I propose to run experiments out of OpenCV tree and remove the test folder from git ignore list.  General compiled files looks ok.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+        std::cerr << "Opening camera..." << std::endl;
+
+    if (getCameraManager()->cameras().size() == 0)
+        throw std::runtime_error("no cameras available");
```
You can use CV_Assert and CV_Error here and bellow. OpenCV CV_ macros also add file name and line number. It's very useful for debug.

QueenofUSSR replied:
Fixed in https://github.com/opencv/opencv/pull/27554/commits/ac4d85e7baf4641f1b2aaace14d789e192b696d7

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+        throw std::runtime_error("failed to acquire camera " + cam_id);
+    camera_acquired_ = true;
+
+    if (options_->verbose)
+        std::cerr << "Acquired camera " << cam_id << std::endl;
```
CV_LOG_ERROR. No need own flag for it. OpenCV already have logging mechanism.

QueenofUSSR replied:
Fixed in https://github.com/opencv/opencv/pull/27554/commits/2d65ed20c982ac658819b5a40a3f74cb8059a99b

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+        throw std::runtime_error("failed to generate viewfinder configuration");
+
+    // Now we get to override any of the default settings from the options_->
+    configuration_->at(0).pixelFormat = libcamera::formats::RGB888;
```
AI to Alexander: check BGR.

asmorkalov commented on modules/videoio/src/cap_libcamera.hpp:
```diff
@@ -0,0 +1,539 @@
+    std::unique_ptr<CameraConfiguration> configuration_;
+    std::map<FrameBuffer *, std::vector<libcamera::Span<uint8_t>>> mapped_buffers_;
+    std::map<std::string, Stream *> streams_;
+    FrameBufferAllocator *allocator_ = nullptr;
```
I propose to convert allocator from pointer to regular filed. No need to call new/delete for it each time. It allows to prevent memory leaks.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+{
+    CV_LOG_DEBUG(NULL, "VIDEOIO(Libcamera:" << _deviceName << "): opening...");
+    // Some parameters initialization here, maybe more needed.
+    options->video_width = 1280;
+    options->video_height = 720;
```
V4L2 uses resolution 640x480 by default as most common. Not all cameras support 720p.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+        return false;
+    }
+
+    // if (needsReconfigure)
+    // {
+    //     if (isFramePending)
+    //     {
+    //         stopVideo();
+    //         startVideo();
+    //     }
+    // }
```
please drop dead code.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+    BufferMap buffers(std::move(completed_request->buffers));
+
+    Request *request = completed_request->request;
+    assert(request);
```
Please use CV_Assert. Plain C assert does not work in release by default. Use CV_Dbg_Assert, if it's debug only check.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+    controls_.set(NoiseReductionMode, denoise);
+}
+
+/* ***************End of methods from original LibcameraApp class*************** */
```
The comment most probably is not relevant any more.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1100 @@
+    current_request_ = nullptr;
+}
+
+LibcameraCapture::LibcameraCapture(int camera_index)
+{
+    // LibcameraCapture();
+    options_ = std::unique_ptr<Options>(new Options());
+    controls_ = controls::controls;
+    if (!options_)
+        options_ = std::make_unique<Options>();
+    controls_.clear();
+    // original constructor ends
+    options = static_cast<Options *>(GetOptions());
+    still_flags = FLAG_STILL_NONE;
+    options->photo_width = 2560;
+    options->photo_height = 1440;
+    options->video_width = 640;
+    options->video_height = 480;
+    options->framerate = 30;
+    options->denoise = "auto";
+    options->timeout = 1000;
+    options->setMetering(Metering_Modes::METERING_MATRIX);
+    options->setExposureMode(Exposure_Modes::EXPOSURE_NORMAL);
+    options->setWhiteBalance(WhiteBalance_Modes::WB_AUTO);
+    options->contrast = 1.0f;
+    options->saturation = 1.0f;
+    options->camera = camera_index;
+    // still_flags |= LibcameraApp::FLAG_STILL_RGB;
+    still_flags |= FLAG_STILL_BGR;
+    needsReconfigure.store(false, std::memory_order_release);
+    camerastarted = false;
+    current_request_ = nullptr;
+    // Automatically open the camera
+    open(camera_index);
+}
```
No need to duplicate camera initialization in the second constructor. You can call the first one and ten call open like here: https://stackoverflow.com/questions/308276/can-i-call-a-constructor-from-another-constructor-do-constructor-chaining-in-c

QueenofUSSR replied:
Fixed in https://github.com/opencv/opencv/pull/27554/commits/7c2d8aa12ff2b893f9da55d374537e8328850abd

asmorkalov commented on modules/videoio/src/cap_libcamera.hpp:
```diff
@@ -0,0 +1,539 @@
+    void StreamDimensions(Stream const *stream, unsigned int *w, unsigned int *h, unsigned int *stride) const;
+
+protected:
+    std::unique_ptr<Options> options_;
```
Why do you need unique_ptr here? Why not just local field?

asmorkalov committed:
Switch CI to container with libcamera-dev.

ssxc16 committed:
Normalize logging: replace std::cerr with CV_LOG_*

Commit:
update .gitignore

Commit:
Fix with CV_* macro

Commit:
Clear dead code and refactor the constructor

Commit:
delete unnecessary comments

Commit:
Replace plain C assert with CV_Assert

Lookplanes committed:
Fix with CV_* macro AND Normalize logging: replace std::cerr with CV_LOG_*

Lookplanes committed:
Backup: Before LibcameraApp removal refactoring

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+void LibcameraCapture::OpenCamera()
+{
+
+    if (options_->verbose)
```
verbose flag is not needed any more. CV_LOG_ messages are controlled by enviroment variables  OpenCV. See logging section of https://docs.opencv.org/4.x/d6/dea/tutorial_env_reference.html

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+
+    // An application could be holding a CompletedRequest while it stops and re-starts
+    // the camera, after which we don't want to queue another request now.
+    {
```
stop_lock(camera_stop_mutex_) is not release here. Is it expected?

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+            CV_Error(cv::Error::StsAssert, "failed to add buffer to request in QueueRequest");
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(control_mutex_);
+        request->controls() = std::move(controls_);
+    }
```
stop_lock(camera_stop_mutex_) is not release here. Is it expected?

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+    controls_.clear();
+    options = static_cast<Options *>(GetOptions());
+    still_flags = FLAG_STILL_NONE;
+    options->photo_width = 2560;
+    options->photo_height = 1440;
```
I propose to set minimal resolution to be sure the VideoCapture opens with any camera. I would say, 640x480 should be enough as for preview.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+    camerastarted = false;
+    current_request_ = nullptr;
+    // Automatically open the camera
+    if (camera_index >= 0) {
+        open(camera_index);
+    }
```
It silently ignores incorrect input. I propose to remove check here and add CV_Assert at the very beginning. No much sense to fill options, if camera index is not correct.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+    catch (const std::exception& e)
+    {
+        // No message available or error occurred
+        return false;
```
Please print exception message with CV_LOG_ERROR here.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1001 @@
+        return false;
+    }
+    
+    // Create OpenCV Mat directly from libcamera buffer
+    Mat frame(vh, vw, CV_8UC3);
+    uint8_t *src_ptr = mem[0].data();
+    uint line_size = vw * 3;
+    
+    // Copy line by line to handle stride correctly
+    for (unsigned int i = 0; i < vh; i++, src_ptr += vstr)
+    {
+        memcpy(frame.ptr(i), src_ptr, line_size);
+    }
+    
+    frame.copyTo(dst);
+    return true;
```
You have 2 copies here. cv::Mat supports stride (step) you can construct cv::Mar from `mem[0].data()` and call copyTo for it.

Commit:
Add libcamera_csi.markdown

Lookplanes committed:
Mega update: Merge  OnDemand improvements into lk_temp

This is a major update that brings all the latest developments from OnDemand branch:
- Zero-copy libcamera implementation optimizations
- Producer-Consumer queue removal for on-demand mode
- Enhanced error handling and logging standardization
- Performance improvements and code cleanup
- All latest OpenCV 4.x upstream updates

This preserves lk_temp branch history while incorporating all new features.

Lookplanes committed:
Merge branch 'lk_temp' of https://github.com/QueenofUSSR/opencv-RPiCamera into lk_temp

Lookplanes committed:
remove verbose and change initial camera resolution to 640x480

asmorkalov commented:
Sorry, missed the previous question. cv::Mat behaves like smart pointer. In case if used saves buffers somewhere both memory allocators: default and cv::Mat base will behave in the same way. We will get  memory high consumption . I do not see any difference here.

asmorkalov commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1316 @@
+LibcameraCapture::LibcameraCapture(int camera_index) : LibcameraCapture()
+{
+    options->camera = camera_index;
+    open(camera_index);
```
return value is ignored.

Lookplanes committed:
Fixed memory leak issue in LibcameraFrameAllocator

Lookplanes committed:
Fix recycle request in copy-fallback path to prevent resource leak

QueenofUSSR commented on modules/videoio/src/cap_libcamera.cpp:
```diff
@@ -0,0 +1,1319 @@
+            return true;
+        } else {
+            // (回退)步长不兼容，使用拷贝
+            CV_LOG_WARNING(NULL, cv::format("Stride mismatch: libcamera=%d, expected=%d", 
+                                           vstr, expected_row_bytes));
```
This is exactly what we are dealing with now. It seems that different devices have diffrent strides.

Commit:
Address issues mentioned above, set aside zero-copy feature temporarily.

Commit:
Merge branch 'lk_new' into lk_temp
